### PR TITLE
core(fr): handle 0 throughput in timespan

### PR DESCRIPTION
**Pull request by adamraine**
Closes https://github.com/GoogleChrome/lighthouse/issues/13321

This appears to be the correct solution given how we calculate savings in timespan mode, but it also means that `wastedMs` will be 0 for every BE audit in timespan mode using the desktop config.

### CODE REVIEW

**Comment by brendankenny** [comment]
I'm on my phone so can't check for myself, sorry, but what's the difference between timespan and navigation for this? I thought we estimated from the download rate in cases like this... Is it we don't have that info in a timespan?

**Comment by adamraine** [comment]
We estimate from the throughput in timespan, which is calculated using `throttling.downloadThroughputKbps` for DevTools throttling (timespan always uses DevTools throttling). `throttling.downloadThroughputKbps` is 0 in the desktop config, this leads to some divide by 0 when calculating `wastedMs`:

https://github.com/GoogleChrome/lighthouse/blob/5d8ea1a4ec52b47bdc62d629ecbdc9f77d4142f0/lighthouse-core/audits/byte-efficiency/byte-efficiency-audit.js#L207-L212

**Comment by brendankenny** [comment]
> but what's the difference between timespan and navigation for this

for myself: the difference is that estimating via throughput isn't done for navigation, that uses the full graph simulation: https://github.com/GoogleChrome/lighthouse/blob/5d8ea1a4ec52b47bdc62d629ecbdc9f77d4142f0/lighthouse-core/audits/byte-efficiency/byte-efficiency-audit.js#L227-L234

If instead we did call `computeWastedMsWithThroughput` in a navigation with `--preset desktop`, we'd have the same `NaN` issue.

> We estimate from the throughput in timespan, which is calculated using `throttling.downloadThroughputKbps` for DevTools throttling

sorry, the missing object in `I thought we estimated from the download rate` was supposed to be "throughput" :)

It looks like we only do that for `'provided'`: https://github.com/GoogleChrome/lighthouse/blob/5d8ea1a4ec52b47bdc62d629ecbdc9f77d4142f0/lighthouse-core/computed/load-simulator.js#L40-L43

which is estimated over in [`network-analyzer`](https://github.com/GoogleChrome/lighthouse/blob/68ba77acbc9c22512b93d213b23b05cca443dda4/lighthouse-core/lib/dependency-graph/simulator/network-analyzer.js#L385) from whatever network records are available.

We know we have network records due to this check:

https://github.com/GoogleChrome/lighthouse/blob/68ba77acbc9c22512b93d213b23b05cca443dda4/lighthouse-core/audits/byte-efficiency/byte-efficiency-audit.js#L120-L128

what if we altered `computeWastedMsWithThroughput` to fallback to the calculated throughput if one wasn't set?

**Comment by brendankenny** [comment]
kind of ugly version of what I mean, ideally we can do this nicer but the file might need a refactor for the FR world it lives in now that we don't want to take on at this point:

```diff
diff --git a/lighthouse-core/audits/byte-efficiency/byte-efficiency-audit.js b/lighthouse-core/audits/byte-efficiency/byte-efficiency-audit.js
index 4c6e17abe..bae49d2ba 100644
--- a/lighthouse-core/audits/byte-efficiency/byte-efficiency-audit.js
+++ b/lighthouse-core/audits/byte-efficiency/byte-efficiency-audit.js
@@ -12,6 +12,7 @@ const i18n = require('../../lib/i18n/i18n.js');
 const NetworkRecords = require('../../computed/network-records.js');
 const LoadSimulator = require('../../computed/load-simulator.js');
 const PageDependencyGraph = require('../../computed/page-dependency-graph.js');
+const NetworkAnalysis = require('../../computed/network-analysis.js');
 
 const str_ = i18n.createMessageInstanceIdFn(__filename, {});
 
@@ -127,16 +128,17 @@ class UnusedBytes extends Audit {
       };
     }
 
-    const [result, graph, simulator] = await Promise.all([
+    const [result, graph, simulator, networkAnalysis] = await Promise.all([
       this.audit_(artifacts, networkRecords, context),
       // Page dependency graph is only used in navigation mode.
       gatherContext.gatherMode === 'navigation' ?
         PageDependencyGraph.request({trace, devtoolsLog}, context) :
         null,
       LoadSimulator.request(simulatorOptions, context),
+      NetworkAnalysis.request(devtoolsLog, context),
     ]);
 
-    return this.createAuditProduct(result, graph, simulator, gatherContext);
+    return this.createAuditProduct(result, graph, simulator, gatherContext, networkAnalysis);
   }
 
   /**
@@ -203,9 +205,13 @@ class UnusedBytes extends Audit {
   /**
    * @param {number} wastedBytes
    * @param {Simulator} simulator
+   * @param {LH.Artifacts.NetworkAnalysis} networkAnalysis
    */
-  static computeWastedMsWithThroughput(wastedBytes, simulator) {
-    const bitsPerSecond = simulator.getOptions().throughput;
+  static computeWastedMsWithThroughput(wastedBytes, simulator, networkAnalysis) {
+    let bitsPerSecond = simulator.getOptions().throughput;
+    if (bitsPerSecond === 0) {
+      bitsPerSecond = networkAnalysis.throughput;
+    }
     const wastedBits = wastedBytes * 8;
     const wastedMs = wastedBits / bitsPerSecond * 1000;
     return wastedMs;
@@ -216,9 +222,10 @@ class UnusedBytes extends Audit {
    * @param {Node|null} graph
    * @param {Simulator} simulator
    * @param {LH.Artifacts['GatherContext']} gatherContext
+   * @param {LH.Artifacts.NetworkAnalysis} networkAnalysis
    * @return {LH.Audit.Product}
    */
-  static createAuditProduct(result, graph, simulator, gatherContext) {
+  static createAuditProduct(result, graph, simulator, gatherContext, networkAnalysis) {
     const results = result.items.sort((itemA, itemB) => itemB.wastedBytes - itemA.wastedBytes);
 
     const wastedBytes = results.reduce((sum, item) => sum + item.wastedBytes, 0);
@@ -230,7 +237,7 @@ class UnusedBytes extends Audit {
         providedWastedBytesByUrl: result.wastedBytesByUrl,
       });
     } else {
-      wastedMs = this.computeWastedMsWithThroughput(wastedBytes, simulator);
+      wastedMs = this.computeWastedMsWithThroughput(wastedBytes, simulator, networkAnalysis);
     }
 
     let displayValue = result.displayValue || '';
```

extra computation cost is minimal because simulator creation also runs `NetworkAnalysis.request` so it'll be cached for one or the other.

> We know we have network records due to this check:

this should cover us, but `networkAnalysis.throughput` also [falls back to `Infinity`](https://github.com/GoogleChrome/lighthouse/blob/68ba77acbc9c22512b93d213b23b05cca443dda4/lighthouse-core/lib/dependency-graph/simulator/network-analyzer.js#L407-L409) if there are no usable network records, which is equivalent to `computeWastedMsWithThroughput` returning 0 as this PR does, so we're covered either way

**Comment by adamraine** [comment]
> If instead we did call computeWastedMsWithThroughput in a navigation with --preset desktop, we'd have the same NaN issue, but we don't do that so we don't :)

The simulator doesn't have any NaN issues with navigations, but does it make sense to have simulator options `throughput` set to 0? My whole confusion here stems from the fact that 0 throughput in the config represents "unset". Should we be setting `throughput` to the network analysis value in *all* cases where config `throughput` is 0?

File: lighthouse-core/computed/load-simulator.js [hunk 1]
```diff
@@ -63,9 +63,14 @@ class LoadSimulator {
         break;
       default:
         // intentionally fallback to simulator defaults
-        break;
+        return new Simulator(options);
     }
 
+    // 0 should be interpreted as "unset" for these values.
+    // Fallback to environment values in this case.
+    if (options.rtt === 0) options.rtt = networkAnalysis.rtt;
```
Comment (adamraine) [inline lighthouse-core/computed/load-simulator.js:71]: I assumed this is true for `rtt` but it shouldn't be necessary to fix the issue at hand.

**Comment by paulirish** [comment]
> * we should seriously consider giving `desktop` a non-zero `downloadThroughputKbps`/`uploadThroughputKbps` in the near future.

One notable difference is... in lantern, throughput and rtt are used "correctly", in that our simulated environment ensures the throughput affects network transmission to the best of our abilities.  However in our devtools throttling, both the rtt and throttling numbers are **additive**. We'd just adding extra latency and constraining throughput beyond what's happening in real life. (put another way... if you wanted to know what rtt of 2ms and throughput of 1gb looks like... lantern would be ~accurate, but w/ devtools your results would be basically indicative of your machine's actual network conditions)

So... given that we have additive throttling in devtools.. it makes me think our BE `computeWastedMsWithThroughput` calculation is currently pretty innacurate for configs assuming devtools throttling..

**Comment by paulirish** [comment]
three possibilities

1. super minimal: return 0 (current state of PR)
2. smallish impact: use networkAnalyzer.throughput in BE's computeWastedMs
3. impact to be determined: change simulator so both 'provided' + 'devtools' throttling use networkSimulator results.
  - we don't have smoketest coverage that'd capture this so we want to verify the impact.

**Comment by paulirish** [comment]
@adamraine if you wanna land this with option 1 (as listed above), i'm only comfortable doing that as a super quick solution to avoid NaN.

options 2 and 3 are the only options that provide "correct" results, so i'd expect we'd followup in the shortterm with one of those.

**Comment by adamraine** [comment]
Lets do option 1 for v9, that's what's implemented now

File: lighthouse-core/audits/byte-efficiency/byte-efficiency-audit.js [hunk 2]
```diff
@@ -206,6 +206,7 @@ class UnusedBytes extends Audit {
    */
   static computeWastedMsWithThroughput(wastedBytes, simulator) {
     const bitsPerSecond = simulator.getOptions().throughput;
+    if (bitsPerSecond === 0) return 0;
```
Comment (connorjclark) [inline lighthouse-core/audits/byte-efficiency/byte-efficiency-audit.js:209]:
```suggestion
    // https://github.com/GoogleChrome/lighthouse/pull/13323#issuecomment-962031709
    if (bitsPerSecond === 0) return 0;

```

**Review by connorjclark — APPROVED** [review]

**Pull request merged**